packages/hardhat/test/e2e/crab-Migration.ts

```diff
@@ -0,0 +1,445 @@
+
+            [, , collatV1, shortV1] = await crabStrategyV1.getVaultDetails();
+            [, , collatV2, shortV2] = await crabStrategyV2.getVaultDetails();
+        }
+
+        const decreaseCR1 = async () => {
+            // Withdraw from crabv1's vault to decrease CR1
+            await provider.send("hardhat_setBalance", [crabStrategyV1.address, ethers.utils.parseEther('10').toHexString()]) // Gas
+            await provider.send("hardhat_impersonateAccount", [crabStrategyV1.address]);
+
+            const signer = ethers.provider.getSigner(crabStrategyV1.address);
+            await controller.connect(signer).withdraw(vaultIdV1, ethers.utils.parseEther("10"))
+
+            await provider.send("hardhat_setBalance", [crabStrategyV1.address, "0x0"])
+            await provider.send('evm_mine', []);
+            await provider.send('hardhat_stopImpersonatingAccount', [crabStrategyV1.address]);
+
+            [, , collatV1, shortV1] = await crabStrategyV1.getVaultDetails();
+            [, , collatV2, shortV2] = await crabStrategyV2.getVaultDetails();
+        }
+
+        it("Should migrate d4 when CR1 > CR2", async () => {
+            await initialize(d4.address)
+            await increaseCR1()
+            const [isFlashMigrate, ethNeededForV2, v1oSqthToPay, ethToGetFromV1] = await crabMigration.flashMigrationDetails(d4.address)
+            const crabV2SharesBefore = await crabStrategyV2.balanceOf(d4.address)
+
+            expect(wdiv(collatV1, shortV1).gt(wdiv(collatV2, shortV2))).to.be.true
+            expect(isFlashMigrate).to.be.true
+            expect(ethToGetFromV1.gt(ethNeededForV2)).to.be.true
+
+            const excessEth = ethToGetFromV1.sub(ethNeededForV2)
+            const ethToFlashDeposit = excessEth.mul(180).div(100) // 1.8 times Something smaller than 2
+
+            await crabStrategyV1.connect(d4).approve(crabMigration.address, crabV1SharesBefore);
+            await expect(crabMigration.connect(d5).flashMigrateAndWithdrawFromV1toV2(one, 0, one)).to.be.revertedWith("M12");
+            await crabMigration.connect(d4).flashMigrateFromV1toV2(0);
+
+            const crabV1SharesAfter = await crabStrategyV1.balanceOf(d4.address)
+            const crabV2SharesAfter = await crabStrategyV2.balanceOf(d4.address)
+            const crabV1SharesInMigration = await crabStrategyV1.balanceOf(crabMigration.address)
+            const crabV2ShraesInMigration = await crabStrategyV2.balanceOf(crabMigration.address)
+            const userEthBalanceAfter = await provider.getBalance(d4.address)
+
+            expect(crabV1SharesAfter.eq(0)).to.be.true
+            expect(crabV1SharesInMigration.eq(0)).to.be.true
+            expect(crabV2SharesAfter.gte(crabV2SharesBefore)).to.be.true // There will lil more crab because of the flash deposit
+            expect(userEthBalanceAfter.gt(userEthBalance)).to.be.true // Unused ETH from flash deposit will be returned
```

Comment:
ideally I think its good to calculate these vs just saying > or <, etc and compare the result with what the user has.

Reply:
Same for crab v2 shares.